### PR TITLE
Add mocks in MakeFile for generating mocks

### PR DESCRIPTION
**Pull request by Rizwana777**
#### Description:
- Added Makefile target for generating mocks

#### Link to JIRA Story (if applicable):
https://issues.redhat.com/browse/GITOPSRVCE-732

### CODE REVIEW

**Review by keithchong — APPROVED** [review]
LGTM, the CI check is passing.

File: backend-shared/Makefile [hunk 1]
```diff
@@ -67,5 +67,10 @@ envtest: ## Download envtest-setup locally if necessary.
 test: fmt vet envtest ## Run tests.
 	ACK_GINKGO_DEPRECATIONS=1.16.4 KUBEBUILDER_ASSETS="$(shell $(ENVTEST) use $(ENVTEST_K8S_VERSION) -p path)" go test -p=1 ./... -coverpkg=./... -coverprofile cover.out
 
+mocks: ##  generate mocks
+	cd $(shell pwd)/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
+	cd $(shell pwd)/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_database_queries.go -package=mocks github.com/redhat-appstudio/managed-gitops/backend-shared/db DatabaseQueries
+	cd $(shell pwd)/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=status-writer.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client StatusWriter
+	cd $(shell pwd)/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=cr-client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
```
Comment (jgwest) [inline backend-shared/Makefile:74]: This fails when run

**Review by jgwest — CHANGES_REQUESTED** [review]
@Rizwana777 @keithchong 

1)

For future reference, for a PR like this, you may want to try the PR before submitting/reviewing it. When I ran it there was an obvious error:
```
[jgw@localhost-lan managed-gitops]$ make mocks
OS is linux
Arch is amd64
cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared && make mocks
make[1]: Entering directory '/home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared'
cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_database_queries.go -package=mocks github.com/redhat-appstudio/managed-gitops/backend-shared/db DatabaseQueries
cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=status-writer.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client StatusWriter
/bin/sh: line 1: cd: too many arguments
make[1]: *** [Makefile:73: mocks] Error 1
make[1]: Leaving directory '/home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared'
make: *** [Makefile:307: mocks] Error 2
```

2) There appear to be other mocked packages that are missing from the PR:
```
[jgw@localhost-lan managed-gitops]$ find . | grep mock
./backend-shared/apis/managed-gitops/v1alpha1/mocks/generate.go
./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs
./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/builders.go
./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/gitopsdeployment.go
./backend-shared/apis/managed-gitops/v1alpha1/mocks/cr-client.go
./backend-shared/apis/managed-gitops/v1alpha1/mocks/status-writer.go

./backend-shared/util/log/mocks/log_sink_mock.go

./backend-shared/util/mocks/mock_client.go
./backend-shared/util/mocks/mock_database_queries.go

./backend/condition/mocks/conditions.go

./cluster-agent/utils/mocks/ApplicationServiceClient.go
./cluster-agent/utils/mocks/Client.go
./cluster-agent/utils/mocks/SessionServiceClient.go
./cluster-agent/utils/mocks/SettingsServiceClient.go
./cluster-agent/utils/mocks/VersionServiceClient.go
```

3) I believe this item is missing from the acceptance criteria is missing from the PR:

> Add a CI check that verifies that the mocks are up to date (similar to the other CI checks we already have, for example, see the CI check that verifies that the db-schema.sql is consistent with the migration steps)

See `git diff --exit-code -- .` in static.yml for an example of this

**Comment by Rizwana777** [comment]
> @Rizwana777 @keithchong
> 
> 1. For future reference, for a PR like this, you may want to try the PR before submitting/reviewing it. When I ran it there was an obvious error:
> 
> ```
> [jgw@localhost-lan managed-gitops]$ make mocks
> OS is linux
> Arch is amd64
> cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared && make mocks
> make[1]: Entering directory '/home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared'
> cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
> cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_database_queries.go -package=mocks github.com/redhat-appstudio/managed-gitops/backend-shared/db DatabaseQueries
> cd /home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=status-writer.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client StatusWriter
> /bin/sh: line 1: cd: too many arguments
> make[1]: *** [Makefile:73: mocks] Error 1
> make[1]: Leaving directory '/home/jgw/workspace/delme/review/rizwana/managed-gitops/backend-shared'
> make: *** [Makefile:307: mocks] Error 2
> ```
> 
> 2. There appear to be other mocked packages that are missing from the PR:
> 
> ```
> [jgw@localhost-lan managed-gitops]$ find . | grep mock
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/generate.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/builders.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/gitopsdeployment.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/cr-client.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/status-writer.go
> 
> ./backend-shared/util/log/mocks/log_sink_mock.go
> 
> ./backend-shared/util/mocks/mock_client.go
> ./backend-shared/util/mocks/mock_database_queries.go
> 
> ./backend/condition/mocks/conditions.go
> 
> ./cluster-agent/utils/mocks/ApplicationServiceClient.go
> ./cluster-agent/utils/mocks/Client.go
> ./cluster-agent/utils/mocks/SessionServiceClient.go
> ./cluster-agent/utils/mocks/SettingsServiceClient.go
> ./cluster-agent/utils/mocks/VersionServiceClient.go
> ```
> 
> 3. I believe this item is missing from the acceptance criteria is missing from the PR:
> 
> > Add a CI check that verifies that the mocks are up to date (similar to the other CI checks we already have, for example, see the CI check that verifies that the db-schema.sql is consistent with the migration steps)
> 
> See `git diff --exit-code -- .` in static.yml for an example of this

@jgwest 1) for 1st one I tried ran make mocks and it is not failing for me 
```
```
rnaaz@rnaaz-mac managed-gitops % make mocks
OS is darwin
Arch is arm64
Mac arm64 detected
cd /Users/rnaaz/Documents/workspace/managed-gitops/backend-shared && make mocks
cd /Users/rnaaz/Documents/workspace/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
cd /Users/rnaaz/Documents/workspace/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_database_queries.go -package=mocks github.com/redhat-appstudio/managed-gitops/backend-shared/db DatabaseQueries
cd /Users/rnaaz/Documents/workspace/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=status-writer.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client StatusWriter
cd /Users/rnaaz/Documents/workspace/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=cr-client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
```
```

**Comment by Rizwana777** [comment]
> @Rizwana777 strange, maybe it's a Mac thing? I'm surprised `cd` command would not throw an error in that case. In the script you posted, it's definitely missing `&&` for those lines.

Yes , I think it is a mac specific issue, `&& mockgen` this is missing in last 2 lines

**Comment by Rizwana777** [comment]
> 
> 2. There appear to be other mocked packages that are missing from the PR:
> 
> ```
> [jgw@localhost-lan managed-gitops]$ find . | grep mock
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/generate.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/builders.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/gitopsdeployment.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/cr-client.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/status-writer.go
> 
> ./backend-shared/util/log/mocks/log_sink_mock.go
> 


> ./backend-shared/util/mocks/mock_client.go
> ./backend-shared/util/mocks/mock_database_queries.go

@jgwest  above mentioned backend-shared mocks are almost covered and others such as these
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/generate.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/builders.go
> ./backend-shared/apis/managed-gitops/v1alpha1/mocks/structs/gitopsdeployment.go
cannot be covered as these are not generated mocks
> 
> ./backend/condition/mocks/conditions.go
Added this in recent commit
> 
> ./cluster-agent/utils/mocks/ApplicationServiceClient.go
> ./cluster-agent/utils/mocks/Client.go
> ./cluster-agent/utils/mocks/SessionServiceClient.go
> ./cluster-agent/utils/mocks/SettingsServiceClient.go
> ./cluster-agent/utils/mocks/VersionServiceClient.go
> ```
> 
covered this as well  in recent changes

**Comment by Rizwana777** [comment]
@jgwest Addressed all the comments, PTAL

**Comment by keithchong** [comment]
Yeah, it's odd.  When I ran it, it was ok:

```
kykchong@kykchong-mac managed-gitops % make mocks                                                                                                                    
OS is darwin
Arch is arm64
Mac arm64 detected
cd /Users/kykchong/wrh/git/managed-gitops/backend-shared && make mocks
cd /Users/kykchong/wrh/git/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
cd /Users/kykchong/wrh/git/managed-gitops/backend-shared/util/mocks && mockgen -build_flags=--mod=mod -destination=mock_database_queries.go -package=mocks github.com/redhat-appstudio/managed-gitops/backend-shared/db DatabaseQueries
cd /Users/kykchong/wrh/git/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=status-writer.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client StatusWriter
cd /Users/kykchong/wrh/git/managed-gitops/backend-shared/apis/managed-gitops/v1alpha1/mocks -build_flags=--mod=mod -destination=cr-client.go -package=mocks sigs.k8s.io/controller-runtime/pkg/client Client
kykchong@kykchong-mac managed-gitops %
```

**Review by jgwest — APPROVED** [review]
Looks perfect, thanks @Rizwana777!

**Pull request merged**